### PR TITLE
feat: add datafusion at 0.4.0

### PR DESCRIPTION
**Pull request by cpcloud**

### CODE REVIEW

File: conda-forge.yml [hunk 1]
```diff
@@ -1,5 +1,6 @@
 channel_priority: strict
```
Comment (xhochy) [inline conda-forge.yml:1]: Don't modify the top-level conda-forge.yml, place one on the recipe dir.

File: recipes/datafusion/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  sha256: f98aa882d78a7cd86d23641b557bffe422169b73dc387eb9dddc0e366e5970bf
+
+build:
+  number: 0
+  script: "{{ PYTHON }} -m pip install . -vv"
+
+requirements:
+  build:
+    - {{ compiler('c') }}
+    - rust
```
Comment (xhochy) [inline recipes/datafusion/meta.yaml:19]:
```suggestion
    - {{ compiler('rust') }}
```

File: recipes/datafusion/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+test:
+  imports:
+    - datafusion
+
+about:
+  home: https://github.com/apache/arrow-datafusion
+  license: Apache-2.0
+  license_family: APACHE
+  license_file: LICENSE.txt
```
Comment (xhochy) [inline recipes/datafusion/meta.yaml:38]: You need to generate a second license file using `cargo-bundle-licenses` that bundle the statically linked lib's licences, see https://github.com/conda-forge/cargo-bundle-licenses-feedstock/blob/18cd199df604fee2f36ca6748a7feb29fc754cf7/recipe/build.sh#L14

File: conda-forge.yml (hunk 1 above)
Comment (cpcloud) [reply conda-forge.yml:1]: Good call. Thanks.

File: recipes/datafusion/meta.yaml (hunk 3 above)
Comment (cpcloud) [reply recipes/datafusion/meta.yaml:38]: This seems like the responsibility of the upstream package, no?

Comment (cpcloud) [reply recipes/datafusion/meta.yaml:38]: Also, is there some documentation on why this is necessary? Do all rust packages on conda-forge currently conform to this practice?

Comment (xhochy) [reply recipes/datafusion/meta.yaml:38]: Not sure, you don't need that in a source distribution, only when you ship static binaries. We've been adding that to all rust packages, haven't seen one yet that does it in upstream :(

Comment (xhochy) [reply recipes/datafusion/meta.yaml:38]: > Also, is there some documentation on why this is necessary? Do all rust packages on conda-forge currently conform to this practice?

Also not sure whether we have that documented somewhere but all rust (and Go, Haskell, …) packages should conform to this practice. We have been enforcing it only in the past 12months, so older ones likely miss it. 

Here's also an example line where `cargo-bundle-licenses` is installed as a conda package: https://github.com/conda-forge/ripgrep-feedstock/blob/ac840d94244c5a7293e71015df9913944ed0ad7b/recipe/build.sh#L6

Comment (cpcloud) [reply recipes/datafusion/meta.yaml:38]: I don't think there are any binaries here, it's a library.

Comment (xhochy) [reply recipes/datafusion/meta.yaml:38]: That also a binary ;) "binary" as a synoym for "compiled-and-statically-linked code", not executable.

Comment (cpcloud) [reply recipes/datafusion/meta.yaml:38]: Can you elaborate on (or link to) why this is necessary? Is there a legal reason, or something else?

Comment (xhochy) [reply recipes/datafusion/meta.yaml:38]: See https://github.com/conda-forge/conda-forge.github.io/issues/1052 for more context on this.

File: recipes/datafusion/meta.yaml (hunk 2 above)
Comment (cpcloud) [reply recipes/datafusion/meta.yaml]: This changes breaks the build.

File: recipes/datafusion/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  sha256: f98aa882d78a7cd86d23641b557bffe422169b73dc387eb9dddc0e366e5970bf
+
+build:
+  number: 0
+  script: "{{ PYTHON }} -m pip install . -vv"
+
+requirements:
+  build:
+    - {{ compiler('c') }}
+    - {{ compiler('rust') }}
```
Comment (xhochy) [inline recipes/datafusion/meta.yaml:19]: Sorry, I accidentially added a macOS space in here.
```suggestion
    - {{ compiler('rust') }}
```

Comment (cpcloud) [reply recipes/datafusion/meta.yaml:19]: No worries! Thanks.

**Comment by cpcloud** [comment]
@xhochy I think this is ready for another review if you have a chance

**Pull request merged**